PBKDF2 Hashing feature + minor (#2935)
wso2/docs-is: en/docs/deploy/secure-userstore-using-pbkdf2.md

# PBKDF2 hashing

> We can use Password-Based Key Derivation Function 2 (PBKDF2) hashing method to securely store user passwords in user stores. This method reduces the risk of brute-force attacks due to insecure passwords.

Reviewer: better to have a sentence about PBKDF2 being a modern password hashing algorithm backed by NIST standards.

Author: +1

> We can use Password-Based Key Derivation Function 2 (PBKDF2) hashing method to securely store user passwords in user stores. This method reduces the risk of brute-force attacks due to insecure passwords.

Author: ```suggestion
Password-Based Key Derivation Function 2 (PBKDF2) hashing algorithm is a modern hashing algorithm recommended by NIST.  

We can use PBKDF2 hashing method to securely store user passwords in user stores. This method reduces the risk of brute-force attacks due to insecure passwords.
```

This guide walks you through the steps of configuring PBKDF2 as the hashing algorithm of a JDBC userstore.

!!! note

>     Currently PBKDF2 supports only JDBC userstores in Asgardeo.

Reviewer: this is wrong

Author: Updated.

>     Currently PBKDF2 supports only JDBC userstores in Asgardeo.

Author: ```suggestion
    Currently, PBKDF2 supports only JDBC userstores in WSO@ Identity Server.
```

## Prerequisite

> You should have a secondary JDBC userstore added.  [Create a JDBC user store](../configure-secondary-user-stores) if you don’t already have one.

Reviewer 2: We cannot do this for an already existing userstore if there are users in it. Because once this is enabled, the existing users will not be able to authenticate. We need to follow this when we are creating a userstore or for an existing userstore without any users.

Author: - Removing the pre-requisite section and adding this content in the instructions with a note.

> You should have a secondary JDBC userstore added.  [Create a JDBC user store](../configure-secondary-user-stores) if you don’t already have one.

Reviewer: As stated in https://github.com/wso2/product-is/issues/12124#issuecomment-910492002 we do support PBKDF2 for primary user store. Better to include this information too. However note that this can be used before the initial server startup only (otherwise the admin password validation will fail).

@divyaamunugama FYI: https://github.com/wso2/product-is/issues/12394#issue-982427664

@UdeshAthukorala you implemented a utility that can be used to update the admin password right? can you provide more details on how this can be achieved.

Reviewer 3: @dewniMW The client[1] that I created is used to generate a PBKDF2 Hash and the Salt Value for a given plain text password. If we know the password we can use this client to generate pbkdf2 hash and the salt value and update database directly.

[1]https://github.com/wso2-extensions/identity-hash-provider-pbkdf2/tree/main/utility/pbkdf2-hash-generator

## Configure PBKDF2 hashing
To configure PBKDF2 hashing on a JDBC user store: 

> 1. On Identity Server console application (`https://<IS_HOST>:<PORT>/console`), go to  **Manage > Userstores**.

Reviewer 2: Are we using the console now? Don't we need to explain this with the management console ?

Author: This is a mistake will update it. Thank you!

> 1. On Identity Server console application (`https://<IS_HOST>:<PORT>/console`), go to  **Manage > Userstores**.

Author: ```suggestion
1. On Identity Server management console (`https://<IS_HOST>:<PORT>`), go to  **Manage > Userstores**.
```

# PBKDF2 hashing

> Password-Based Key Derivation Function 2 (PBKDF2) hashing algorithm is a modern hashing algorithm recommended by NIST.  We can use PBKDF2 hashing method to securely store user passwords in user stores. This method reduces the risk of brute-force attacks due to insecure passwords.

Reviewer 2: Check whether we have two spaces here > `NIST.  We can`

Author: ```suggestion
Password-Based Key Derivation Function 2 (PBKDF2) hashing algorithm is a modern hashing algorithm recommended by NIST. We can use the PBKDF2 hashing method to securely store user passwords in user stores. This method reduces the risk of brute-force attacks due to insecure passwords.
```

This guide walks you through the steps of configuring PBKDF2 as the hashing algorithm of a JDBC userstore.

!!! note
    Currently, PBKDF2 supports only JDBC userstores of WSO2 Identity Server.

## Configure PBKDF2 hashing
To configure PBKDF2 hashing on a JDBC user store:

> 1. On Identity Server management console (`https://<IS_HOST>:<PORT>`) and [create a JDBC user store](../configure-secondary-user-stores).

Reviewer 2: Console URL is wrong here, this just has the host and the port.

Reviewer 2: Don't we need to say `Login in to the Identity server....`?

Author: ```suggestion
1. Login to the Identity Server management console (`https://<IS_HOST>:<PORT>/console`) and [create a JDBC user store](../configure-secondary-user-stores).
```

    !!! info

>         - You may also use an existing user store which does not have any users in it. If you already have users in the userstore, once the hashing algorithm is configured these users will not be able to get authenticated.

Reviewer 2: Better to tell how we do this for the existing userstore. This will be the mostly used case.

>         - PBKDF2 is also supported by [primary JDBC userstores](../configure-a-jdbc-user-store), but PBKDF2 should be enabled before the initial server startup by adding the following to the `deployment.toml` file.

Reviewer 2: Shall we highlight this? `should be enabled before the initial server startup`

Author: ```suggestion
        - PBKDF2 is also supported by [primary JDBC userstores](../configure-a-jdbc-user-store), but PBKDF2 **should be enabled before the initial server startup** by adding the following to the `deployment.toml` file.
```

# PBKDF2 hashing
Password-Based Key Derivation Function 2 (PBKDF2) hashing algorithm is a modern hashing algorithm recommended by NIST. We can use the PBKDF2 hashing method to securely store user passwords in user stores. This method reduces the risk of brute-force attacks due to insecure passwords.

This guide walks you through the steps of configuring PBKDF2 as the hashing algorithm of a JDBC userstore.

!!! note
    Currently, PBKDF2 supports only JDBC userstores of WSO2 Identity Server.

## Configure PBKDF2 hashing

> To configure PBKDF2 hashing on a JDBC user store:

Reviewer: ```suggestion
To configure PBKDF2 hashing on a secondary JDBC user store:
```

    !!! info "Existing userstores"
        - You may also use an existing user store which does not have any users in it. If you already have users in the userstore, once the hashing algorithm is configured these users will not be able to get authenticated.
        - Such cases will impact with bad user experience as the users will not get authenticated even when they try to login using the correct credentials. Admins may use the following approaches to reset the user passwords after configuring the PBKDF2 hashing algorithm on an existing userstore:
            - Ask users to [reset their own passwords](../guides/my-account/my-account.md).
            - Trigger password reset for all accounts of the userstore using [admin initiated password reset](../guides/password-mgt/forced-password-reset.md).
    
    !!! info "PBKDF2 for primary JDBC userstores"
        - PBKDF2 is also supported by [primary JDBC userstores](../configure-a-jdbc-user-store), but PBKDF2 **should be enabled before the initial server startup** by adding the following to the `deployment.toml` file.
        ``` js
        [user_store]
        type = "database_unique_id"
        password_digest="PBKDF2"

>         ```

Reviewer: shall we have this as a separate section?
1 sub topic explains the secondary user store steps then the other topic will have the instructions to be followed for the primary user store

Author: Resolved

Added two new L3 heading under [L1] Configure PBKDF2 Hashing.
![image](https://user-images.githubusercontent.com/84217321/170926768-538195f6-afa0-4d3b-882b-82a20d379671.png)

        </tr>
    </table>

5. Click **Update** to save the configurations.

Successful updation of these configurations will convert the password hashing algorithm of the userstore to PBKDF2.

## PBKDF2 parameters
When configuring the PBKDF2 hashing algorithm the following parameters must be specified in the configurations:

<table>
    <tr>
        <th>Parameter</th>
        <th>Parameter name</th>
        <th>Recommended Value</th>
        <th>Description</th>
    </tr>
    <tr>
        <td><code>pbkdf2.iteration.count</code></td>
        <td>Iteration count</td>
        <td><code>10000</code></td>
        <td>Number of times hashing is performed.</td>
    </tr>
        <tr>
        <td><code>pbkdf2.dkLength</code></td>
        <td>Derived Key Length</td>
        <td><code>256</code></td>
        <td>Bit length of the generated hash value.</td>
    </tr>
        <tr>
        <td><code>pbkdf2.prf</code></td>
        <td>Pseudo-Random Function </td>
        <td><code>PBKDF2WithHmacSHA256</code></td>
        <td>The key component of the PBKDF2 hashing algorithm in which the actual hashing part is done.</td>
</table>

!!! info

>     NIST recommends ```PBKDF2WithHmacSHA256``` as the pseudo-random function (prf) value, but the pfr can also be changed. Some examples of possible prf values are as follows:

Reviewer: ```suggestion
    NIST recommends ```PBKDF2WithHmacSHA256``` as the pseudo-random function (prf) value, but the prf can also be changed. Some examples of possible prf values are as follows:
```
